prototype pace sidetransport updates

iskettaneh opened the pull request:
DO NOT REVIEW

Before:
When we publish a new message, we broadcast() on a conditional variable to wake up all waiting goroutines, which causes overhead on the system if the number of waiters is large.

After:
When we publish a new message, we Signal (instead of broadcast) in a paced way up to the number of waiting goroutines.
We maintain two conditional variables, and to waiter counters to signal to ALL waiters that were waiting

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/154046)
<!-- Reviewable:end -->

iskettaneh committed:
prototype: pace closedTS messages

Before:
When we publish a new message, we broadcast() on a conditional variable to wake up all waiting goroutines, which causes overhead on the system if the number of waiters is large.

After:
When we publish a new message, we Signal (instead of broadcast) in a paced way up to the number of waiting goroutines.
We maintain two conditional variables, and to waiter counters to signal to ALL waiters that were waiting

iskettaneh committed:
takepacer: move task pacer to its own util package

Before this commit, the taskPacer lived inside kvserver, which made
it unusable from packages that cannot include the kvserver package.

iskettaneh committed:
Link the sidetransport sender with the task pacer

iskettaneh committed:
Make the task pacing configurable
